lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -960,12 +978,67 @@ void Inliner::inlineInstances(FModuleOp parent) {
   }
 }
 
+void Inliner::identifyNLAsTargetingOnlyModules() {
+  DenseSet<FModuleOp> nlaTargetedModules;
+
+  // Identify candidate NLA's: those that end in a module
+  for (auto &[sym, mnla] : nlaMap) {
+    auto nla = mnla.getNLA();
+    if (nla.isModule())
+      nlaTargetedModules.insert(symbolTable.lookup<FModuleOp>(nla.leafMod()));
+  }
+
+  // Helper to scan leaf modules for users of NLAs, gathering by symbol names
+  auto scanForNLARefs = [&](FModuleOp mod) {
+    DenseSet<StringAttr> referencedNLASyms;
+    // Scan ports
+    for (unsigned i = 0, e = mod.getNumPorts(); i != e; ++i)
+      for (auto anno : AnnotationSet::forPort(mod, i))
+        if (auto sym = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal"))
+          referencedNLASyms.insert(sym.getAttr());
+
+    // Scan operations (and not the module itself):
+    mod.getBody()->walk([&](Operation *op) {
+      for (auto anno : AnnotationSet(op))
+        if (auto sym = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal"))
+          referencedNLASyms.insert(sym.getAttr());
+    });
+
+    return referencedNLASyms;
+  };
+
+  // Walk modules in parallel, scanning for references to NLA's
+  // Gather set of NLA's referenced by each module's ports/operations.
+
+  // Vector of modules and vector for storing results of each walk:
+  SmallVector<FModuleOp, 16> mods(nlaTargetedModules.begin(),
+                                  nlaTargetedModules.end());
+  SmallVector<DenseSet<StringAttr>, 16> refSets;
+  refSets.resize(nlaTargetedModules.size());
+  mlir::parallelForEachN(
+      circuit->getContext(), 0, nlaTargetedModules.size(),
+      [&](size_t index) { refSets[index] = scanForNLARefs(mods[index]); });
+
+  // Combine sets into single "is referenced by a port/operation" set
+  DenseSet<StringAttr> nonModOnlyNLAs;
+  for (auto &set : refSets)
+    nonModOnlyNLAs.insert(set.begin(), set.end());
```

Comment:
You could consider using a `parallelTransformReduce` instead of `parallelForEachN` with a reduction after.